### ISSUE TITLE
[Bug] Structure transformation

### ISSUE THREAD

**Issue by maxeeem** [bowen-xu/PyNARS#47]
**Describe the bug**
Structure transformation is currently not handled correctly.

See this test case from OpenNARS
```
        'Structure transformation

        'Bright is similar to smart. 
        <bright <-> smart>. %0.90;0.90% 

        'Is bright thing a type of smart thing?
        <[smart] --> [bright]>?

        6

        'Bright thing is a type of smart thing. 
        ''outputMustContain('<[bright] <-> [smart]>. %0.90;0.90%')
        ''outputMustContain('<[smart] --> [bright]>. %0.90;0.66%')
```
**To Reproduce**
Launch console `python -m pynars.Console` and type
```
<bright <-> smart>. %0.90;0.90%
<[smart] --> [bright]>?
6
```

**Expected behavior**
The system should be able to derive the following statements:
```
<[bright] <-> [smart]>. %0.90;0.90%
<[smart] --> [bright]>. %0.90;0.66%
```

**Comment by bowen-xu** [comment]
The following rules are missing
```
<S <-> B>.
[S] 
|-
<[S] <-> [P]>.
```

```
<S <-> P>.
{S} 
|-
<{S} <-> {P}>.
```

```
<A <-> B>.
|-
<A --> B>.
<B --> A>.
```

**Comment by maxeeem** [comment]
@bowen-xu I'm not sure additional rules are necessary. This should be handled by structural inference by applying the theorems. We already have some of them https://github.com/bowen-xu/PyNARS/blob/ebff7b965b834334b3612f5431bad20057b3f4cf/pynars/NAL/Theorems/StructuralRules.py#L979

So perhaps just need to check that all of them are in place and that control mechanism correctly applies them.

**Comment by bowen-xu** [comment]
> @bowen-xu I'm not sure additional rules are necessary. This should be handled by structural inference by applying the theorems. We already have some of them
> 
> https://github.com/bowen-xu/PyNARS/blob/ebff7b965b834334b3612f5431bad20057b3f4cf/pynars/NAL/Theorems/StructuralRules.py#L979
> 
> So perhaps just need to check that all of them are in place and that control mechanism correctly applies them.

Yes, the functions have alreadly been there. So we just need to link them to the rule table. See in pr https://github.com/bowen-xu/PyNARS/pull/81

https://github.com/bowen-xu/PyNARS/blob/4000e823388c2dd48cc1b4e1e2e50982b4f13f26/pynars/NARS/InferenceEngine/GeneralEngine/Rules/NAL2.py#L203-L240

**Comment by maxeeem** [comment]
@bowen-xu I guess maybe I'm a little lost and not able to see the connection between the functions. When is the theorem used? I don't see it referenced anywhere if I do a search on the whole codebase, for example `equivalence_theorem4`. Can you please explain how theorems and rules are connected?

And in the new rules, where does the second premise come from? `[S]` or `{S}` for example?

**Comment by bowen-xu** [comment]
@maxeeem I thought `equivalence_theorem4` is a special case of `implication_theorem5`, so that it calls function `_structural__bi_composition__0` when meeting `{<S <-> P>., [S]}` (as 3.0.4 does). In `implication_theorem5` it calls function `bi_composition`, while in `equivalence_theorem4` it can do the same thing (although currently it doesn't call `bi_composition` but is implemented by seperated code).

But maybe you're right; I should make the code more clear by directly calling `equivalence_theorem4` for example.

**Comment by bowen-xu** [comment]
`[S]` is the name of concept `[S]`. It would be passed to the inference engine when there is no beliefs in the concept's _beliefs-table_.
